src/Transports.AspNetCore/WebSockets/BaseSubscriptionServer.cs

```diff
@@ -0,0 +1,488 @@
+                await OnSendKeepAliveAsync();
+            }
+        }
+
+        /*
+         * This 'smart' keep-alive logic doesn't work with subscription-transport-ws because the client will
+         * terminate a connection 20 seconds after the last keep-alive was received (or it will never
+         * terminate if no keep-alive was ever received).
+         * 
+         * See: https://github.com/graphql/graphql-playground/issues/1247
+         */
+        async Task StartSmartKeepAliveLoopAsync()
+        {
+            var lastKeepAliveSent = Connection.LastMessageSentAt;
+            while (true)
+            {
+                var lastSent = Connection.LastMessageSentAt;
+                var lastComm = lastKeepAliveSent > lastSent ? lastKeepAliveSent : lastSent;
+                var now = DateTime.UtcNow;
+                var timePassed = now.Subtract(lastComm);
+                if (timePassed >= keepAliveTimeout)
+                {
+                    await OnSendKeepAliveAsync();
+                    lastKeepAliveSent = now;
+                    await Task.Delay(keepAliveTimeout, CancellationToken);
+                }
+                else
+                {
+                    var timeToWait = keepAliveTimeout - timePassed;
+                    await Task.Delay(timeToWait, CancellationToken);
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Executes when a keep-alive message needs to be sent.
+    /// This method may execute concurrently with other code in this class.
+    /// </summary>
+    protected abstract Task OnSendKeepAliveAsync();
+
+    /// <summary>
+    /// Executes when a connection request needs to be acknowledged.
+    /// </summary>
+    protected abstract Task OnConnectionAcknowledgeAsync(OperationMessage message);
+
+    /// <summary>
+    /// Executes when a new subscription request has occurred.
+    /// Optionally disconnects any existing subscription associated with the same id.
+    /// </summary>
+    protected virtual async Task SubscribeAsync(OperationMessage message, bool overwrite)
+    {
+        if (string.IsNullOrEmpty(message.Id))
+        {
+            await ErrorIdCannotBeBlankAsync(message);
+            return;
+        }
+        var messageId = message.Id!;
+
+        var dummyDisposer = new DummyDisposer();
+
+        try
+        {
+            if (overwrite)
+            {
+                Subscriptions[messageId] = dummyDisposer;
+            }
+            else
+            {
+                if (!Subscriptions.TryAdd(messageId, dummyDisposer))
+                {
+                    await ErrorIdAlreadyExistsAsync(message);
+                    return;
+                }
+            }
+
+            var result = await ExecuteRequestAsync(message);
+            CancellationToken.ThrowIfCancellationRequested();
+            if (!Subscriptions.Contains(messageId, dummyDisposer))
+                return;
+            if (result.Streams?.Count == 1)
```

Comment:
There may be more than 1 stream?

Reply:
There is no support within the GraphQL spec for multiple streams to be returned.  GraphQL.NET has validation rules to prevent this, but I think the execution strategy still allows for it (if the validation rule were not used).  Regardless, the protocol does not support such a scenario as there would need to be separate IDs for each stream.  So, if there were to be more than one stream returned, execution would fall through to SendErrorResultAsync -- but so long as the validation rules are in place, that should not be possible.